### PR TITLE
[DI] Removes number of elements information in debug mode

### PR DESCRIPTION
**Pull request by jschaedl**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31340
| License       | MIT
| Doc PR        | -

With this services config:

```yaml
my_service:
    class: stdClass
    arguments: [!tagged my_tag]

my_tagged_service_1:
    class: stdClass
    tags: [my_tag]

my_tagged_service_2:
    class: stdClass
    tags: [my_tag]
```
Executing `./bin/console debug:container my_service --show-arguments --env=dev` resulted in 

```bash
Information for Service "my_service"
====================================

 ---------------- -------------------------
  Option           Value
 ---------------- -------------------------
  Service ID       my_service
  Class            stdClass
  Tags             -
  Public           no
  Synthetic        no
  Lazy             no
  Shared           yes
  Abstract         no
  Autowired        yes
  Autoconfigured   yes
  Arguments        Iterator (0 element(s))
 ---------------- -------------------------
```
 With this fix the output changed to:

```bash
Information for Service "my_service"
====================================

 ---------------- ------------
  Option           Value
 ---------------- ------------
  Service ID       my_service
  Class            stdClass
  Tags             -
  Public           no
  Synthetic        no
  Lazy             no
  Shared           yes
  Abstract         no
  Autowired        yes
  Autoconfigured   yes
  Arguments        Tagged Iterator for "my_tag"
 ---------------- ------------
```

and with `./bin/console debug:container my_service --show-arguments --env=prod`

```bash
Information for Service "my_service_tagged_iterator"
====================================================

 ---------------- ---------------------------------------------
  Option           Value
 ---------------- ---------------------------------------------
  Service ID       my_service
  Class            stdClass
  Tags             -
  Public           no
  Synthetic        no
  Lazy             no
  Shared           yes
  Abstract         no
  Autowired        yes
  Autoconfigured   yes
  Arguments        Tagged Iterator for "my_tag" (2 element(s))
 ---------------- ---------------------------------------------
```

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php [hunk 1]
```diff
@@ -339,7 +340,11 @@ protected function describeContainerDefinition(Definition $definition, array $op
                 if ($argument instanceof Reference) {
                     $argumentsInformation[] = sprintf('Service(%s)', (string) $argument);
                 } elseif ($argument instanceof IteratorArgument) {
-                    $argumentsInformation[] = sprintf('Iterator (%d element(s))', \count($argument->getValues()));
+                    if ($argument instanceof TaggedIteratorArgument) {
+                        $argumentsInformation[] = sprintf('Tagged Iterator for "%s"', $argument->getTag());
+                    } else {
+                        $argumentsInformation[] = empty($argument->getValues()) ? 'Iterator' : sprintf('Iterator (%d element(s))', \count($argument->getValues()));
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: i'd still expect if we show `0 elements` if the iterator contains 0 elements :)

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php [hunk 2]
```diff
@@ -339,7 +340,11 @@ protected function describeContainerDefinition(Definition $definition, array $op
                 if ($argument instanceof Reference) {
                     $argumentsInformation[] = sprintf('Service(%s)', (string) $argument);
                 } elseif ($argument instanceof IteratorArgument) {
-                    $argumentsInformation[] = sprintf('Iterator (%d element(s))', \count($argument->getValues()));
+                    if ($argument instanceof TaggedIteratorArgument) {
+                        $argumentsInformation[] = sprintf('Tagged Iterator for "%s" %s', $argument->getTag(), !$options['is_debug'] ? sprintf('(%d element(s))', \count($argument->getValues())) : '');
+                    } else {
+                        $argumentsInformation[] = $options['is_debug'] ? 'Iterator' : sprintf('Iterator (%d element(s))', \count($argument->getValues()));
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: We maybe want to be consistent wether we handle the debug or the non-debug case first.

Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: adjusted

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php [hunk 3]
```diff
@@ -337,9 +338,13 @@ protected function describeContainerDefinition(Definition $definition, array $op
                     $argument = $argument->getValues()[0];
                 }
                 if ($argument instanceof Reference) {
-                    $argumentsInformation[] = sprintf('Service(%s)', (string) $argument);
+                    $argumentsInformation[] = sprintf('Service (%s)', (string) $argument);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: this is unrelated and should be reverted - it can break scripts that parse the output

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php [hunk 4]
```diff
@@ -337,9 +338,13 @@ protected function describeContainerDefinition(Definition $definition, array $op
...
                 } elseif ($argument instanceof IteratorArgument) {
-                    $argumentsInformation[] = sprintf('Iterator (%d element(s))', \count($argument->getValues()));
+                    if ($argument instanceof TaggedIteratorArgument) {
+                        $argumentsInformation[] = sprintf('Tagged Iterator for "%s"%s', $argument->getTag(), $options['is_debug'] ? '': sprintf(' (%d element(s))', \count($argument->getValues())));
+                    } else {
+                        $argumentsInformation[] = sprintf('Iterator%s', $options['is_debug'] ? '' : sprintf(' (%d element(s))', \count($argument->getValues())));
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: do we really need the ternary on this line?

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: displaying useful info is more important than "consistency" here to me

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: displaying useful info is more important than "consistency" here to me

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php (hunk 3 above)
Comment (jschaedl) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]:
```suggestion
                    $argumentsInformation[] = sprintf('Service(%s)', (string) $argument);
```

File: src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Descriptor/definition_arguments_1.txt [hunk 5]
```diff
@@ -13,7 +13,7 @@
   Autoconfigured   no                           
   Factory Class    Full\Qualified\FactoryClass  
   Factory Method   get                          
-[39;49m  Arguments        Service(definition2)[39;49m[39;49m         [39;49m
+[39;49m  Arguments        Service (definition2)[39;49m[39;49m        [39;49m
```
Comment (jschaedl) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Descriptor/definition_arguments_1.txt]:
```suggestion
[39;49m  Arguments        Service(definition2)[39;49m[39;49m         [39;49m
```

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php (hunk 3 above)
Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: Hopefully, such script should use the json or XML output, but yes, I still vote for reverting.

Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: reverted

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php (hunk 4 above)
Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: You are right we don't need the ternary. The array of services given for the Iterator is always filled. Will revert it.

Comment (jschaedl) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: done

**Review by Tobion — APPROVED** [review]
We need to be careful when merging this to master due to #31353.
The is_debug option to the can be remove again I guess.

**Comment by nicolas-grekas** [comment]
Thank you @jschaedl.
